File: apps/web/lib/meta-insights-golden.test.ts
// 골든 픽스처 — Meta 응답 파싱을 잠근다 (단계 6).
//
// 게재(요청 바디)와 달리 여기 위험은 **응답 파싱**이다. ad study 결과는 버전·계정마다 모양 편차가 커서
// 파서가 일부러 관대하고(셀 이름 정규화·confidence/p_value 폴백), 광고별 KPI 는 반올림 규칙이 계약이다.
// 그 관대함과 반올림을 Java 재작성본이 똑같이 재현하는지 같은 파일로 지킨다.
//
// 원본(raw)은 손으로 적은 입력이고, 기대값(expected)은 이 클라이언트에서 떠낸 것이다.
//
// 대응 Java: apps/api/src/test/java/ai/adflow/api/meta/MetaInsightsGoldenTest.java

import { readFileSync } from "node:fs";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { metaAdsInsights } from "./meta-ads-insights";

type StudyCase = { name: string; raw: unknown; expected: { winner: "A" | "B" | null; confidence: number } | null };
type AdCase = { name: string; adIds: [string, string]; raw: unknown; expected: unknown };

const FIXTURE = join(__dirname, "../../../packages/contracts/fixtures/meta/insights.json");
const { adStudy, adInsights } = JSON.parse(readFileSync(FIXTURE, "utf8")) as {
  adStudy: StudyCase[];
  adInsights: AdCase[];
};

const fetchMock = vi.fn();

beforeEach(() => {
  fetchMock.mockReset();
  vi.stubGlobal("fetch", fetchMock);
});
afterEach(() => vi.unstubAllGlobals());

describe("ad study 결과 파싱 골든 픽스처", () => {
  it("케이스가 비어 있지 않다", () => {
    expect(adStudy.length).toBeGreaterThanOrEqual(10);
  });

  for (const c of adStudy) {
    it(c.name, async () => {
      fetchMock.mockResolvedValueOnce({ ok: true, json: async () => c.raw });
      expect(await metaAdsInsights.getSplitTestResult("s1", "TOKEN")).toEqual(c.expected);
    });
  }
});

describe("광고별 KPI 파싱 골든 픽스처", () => {
  it("케이스가 비어 있지 않다", () => {
    expect(adInsights.length).toBeGreaterThanOrEqual(5);
  });

  for (const c of adInsights) {
    it(c.name, async () => {
      fetchMock.mockResolvedValueOnce({ ok: true, json: async () => ({ data: [] }) });
      fetchMock.mockResolvedValueOnce({ ok: true, json: async () => c.raw });
      const insights = await metaAdsInsights.getInsights("camp_1", "TOKEN", "all", undefined, "traffic", c.adIds);
      expect(insights.ads).toEqual(c.expected);
    });
  }
});

File: apps/api/src/test/java/ai/adflow/api/meta/MetaInsightsGoldenTest.java
package ai.adflow.api.meta;

import static org.assertj.core.api.Assertions.assertThat;

import ai.adflow.api.tournament.engine.AdKpi;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;

/**
 * 골든 픽스처 — Meta 응답 파싱이 TS 와 같은지 지킨다 (단계 6).
 *
 * <p>대응 TS: apps/web/lib/meta-insights-golden.test.ts
 *
 * <p>원본(raw)은 손으로 적은 입력이고 기대값(expected)은 TS 클라이언트에서 떠낸 것이다. ad study 결과
 * 파서의 <b>관대함</b>(셀 이름 정규화·confidence/p_value 폴백)과 KPI 반올림이 이 파일의 계약이다.
 */
class MetaInsightsGoldenTest {

  private static final ObjectMapper JSON = new ObjectMapper();

  /** 전송을 가로채 미리 정한 응답을 돌려준다. */
  static class StubGraphClient extends MetaGraphClient {
    private final String response;

    StubGraphClient(String response) {
      super(JSON, MetaGraphClient.GRAPH);
      this.response = response;
    }

    @Override
    protected String send(String method, String path, Map<String, Object> body) {
      return response;
    }
  }

  private static JsonNode fixture() {
    try (InputStream in = MetaInsightsGoldenTest.class.getResourceAsStream("/meta/insights.json")) {
      if (in == null) throw new IllegalStateException("픽스처가 클래스패스에 없어요: /meta/insights.json");
      return JSON.readTree(in);
    } catch (java.io.IOException e) {
      throw new IllegalStateException(e);
    }
  }

  @Test
  void 픽스처가_비어_있지_않다() {
    assertThat(fixture().get("adStudy").size()).isGreaterThanOrEqualTo(10);
    assertThat(fixture().get("adInsights").size()).isGreaterThanOrEqualTo(5);
  }

  @TestFactory
  List<DynamicTest> ad_study_결과_파싱이_TS_와_같다() {
    List<DynamicTest> tests = new ArrayList<>();
    for (JsonNode c : fixture().get("adStudy")) {
      tests.add(
          DynamicTest.dynamicTest(
              c.get("name").asString(),
              () -> {
                MetaInsightsClient client =
                    new MetaInsightsClient(new StubGraphClient(JSON.writeValueAsString(c.get("raw"))));
                MetaInsightsClient.StudyResult got = client.splitTestResult("s1", "TOKEN");

                JsonNode want = c.get("expected");
                if (want == null || want.isNull()) {
                  assertThat(got).isNull();
                  return;
                }
                assertThat(got).isNotNull();
                JsonNode wantWinner = want.get("winner");
                assertThat(got.winner())
                    .isEqualTo(wantWinner == null || wantWinner.isNull() ? null : wantWinner.asString());
                assertThat(got.confidence()).isEqualTo(want.get("confidence").asDouble());
              }));
    }
    return tests;
  }

  @TestFactory
  List<DynamicTest> 광고별_KPI_파싱이_TS_와_같다() {
    List<DynamicTest> tests = new ArrayList<>();
    for (JsonNode c : fixture().get("adInsights")) {
      tests.add(
          DynamicTest.dynamicTest(
              c.get("name").asString(),
              () -> {
                MetaInsightsClient client =
                    new MetaInsightsClient(new StubGraphClient(JSON.writeValueAsString(c.get("raw"))));
                List<String> adIds = new ArrayList<>();
                for (JsonNode id : c.get("adIds")) adIds.add(id.asString());

                List<AdKpi> got = client.roundAdKpis("camp_1", "TOKEN", adIds);

                JsonNode want = c.get("expected");
                assertThat(got).hasSize(want.size());
                for (int i = 0; i < want.size(); i++) {
                  JsonNode w = want.get(i);
                  assertThat(got.get(i).impressions()).isEqualTo(w.get("impressions").asInt());
                  assertThat(got.get(i).clicks()).isEqualTo(w.get("clicks").asInt());
                  assertThat(got.get(i).ctr()).isEqualTo(w.get("ctr").asDouble());
                  assertThat(got.get(i).spend()).isEqualTo(w.get("spend").asDouble());
                }
              }));
    }
    return tests;
  }
}
